### PR TITLE
add contributing guide

### PR DESCRIPTION
**Pull request by petrkotas**
For the contributing guide I have added the following:
* I created the recorded demos folder in shared drive https://drive.google.com/drive/folders/1RB1L2-nGMXwsOAOYC-VGGbB0yD3Ae-rD?usp=drive_link
* I updated the contributing guide in the main repo (in the PR)
* I also set the PR template to highlight contribution (in the PR)

Jira: https://issues.redhat.com/browse/ARO-5975

### CODE REVIEW

File: .github/workflows/golangci-lint.yaml
Comment (mjlshen) [inline .github/workflows/golangci-lint.yaml:1]: CI is already present in this repo: https://github.com/Azure/ARO-HCP/blob/main/.github/workflows/ci-go.yml I think that should be modified if needed as opposed to adding a new Action

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+- [ ] PR: The PR description is expressive enough and will help future contributors
+- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
+- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
+- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
+- [ ] Deployment: The deployemnt process was considered and addressed if required
+- [ ] Testing: New code requires new unit tests.
+- [ ] Documentation: Is the documentation updated? Either in the doc located in focus area, in the README or in the code itself.
+- [ ] Customers: Is this change affecting customers? Is the release plan considered?
+
+### Release note
```
Comment (mjlshen) [inline .github/PULL_REQUEST_TEMPLATE.md:43]: Do we use this field in a release process?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+Jira: <!-- optional: link to Jira issue -->
+Link to demo recording: <!-- optional: link to a demo recording -->
+
+### Breadcrumbs
+The following tradeoffs were made:
+
+The following alternatives were considered:
+
+Links to places where the discussion took place: <!-- optional: slack, GH issue, mailinglist, ... -->
```
Comment (mjlshen) [inline .github/PULL_REQUEST_TEMPLATE.md:22]: I think this discussion should be captured in JIRA
```suggestion
```

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (mjlshen) [inline .github/PULL_REQUEST_TEMPLATE.md:1]: I think this is too long, I don't want to fill this out for every PR. Any reason to not use the same one from ARO-RP? https://github.com/Azure/ARO-RP/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
+Approvers are expected to review this list.
+
+- [ ] Design: An [ADR](https://drive.google.com/drive/folders/1zr2gubRU-asXjZSFZvKmcuGOV4jlUGV-) or [DDR](https://drive.google.com/drive/folders/1L9R0mtcl-5oRc9dzRdcw4ZGVJLtj-14o?usp=drive_link) was considered and is present (link) or not required
+- [ ] PR: The PR description is expressive enough and will help future contributors
+- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
+- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
+- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
+- [ ] Deployment: The deployemnt process was considered and addressed if required
```
Comment (mjlshen) [inline .github/PULL_REQUEST_TEMPLATE.md:38]:
```suggestion
- [ ] Deployment: The deployment process was considered and addressed if required
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+### Special notes for your reviewer
+
+<!-- optional -->
+
+### Checklist
+
+This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
+Approvers are expected to review this list.
+
+- [ ] Design: An [ADR](https://drive.google.com/drive/folders/1zr2gubRU-asXjZSFZvKmcuGOV4jlUGV-) or [DDR](https://drive.google.com/drive/folders/1L9R0mtcl-5oRc9dzRdcw4ZGVJLtj-14o?usp=drive_link) was considered and is present (link) or not required
```
Comment (mjlshen) [inline .github/PULL_REQUEST_TEMPLATE.md:33]:
```suggestion
```

**Review by mjlshen — CHANGES_REQUESTED** [review]

File: .github/workflows/ci-go.yml
Comment (mjlshen) [inline .github/workflows/ci-go.yml:1]: Could be convinced, but not seeing the benefit of the matrix style yet. Wouldn't we have to maintain this list by hand if this grows?

File: .github/workflows/golangci-lint.yaml
Comment (petrkotas) [reply .github/workflows/golangci-lint.yaml:1]: Done, my apology for missing this. I did rebased last minute and did not notice. Now the functionality is merged.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (petrkotas) [reply .github/PULL_REQUEST_TEMPLATE.md:22]: Good point, updated.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (petrkotas) [reply .github/PULL_REQUEST_TEMPLATE.md:1]: I am trying to improve the existing process by making it easy to follow checks and process. We learned that remembering everything is hard and missed documentation and decision tracking. Therefore I am providing a bit more guide in the template.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 4 above)
Comment (petrkotas) [reply .github/PULL_REQUEST_TEMPLATE.md:33]: Good catch.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 3 above)
Comment (petrkotas) [reply .github/PULL_REQUEST_TEMPLATE.md:38]: Thanks

File: .github/workflows/ci-go.yml
Comment (petrkotas) [reply .github/workflows/ci-go.yml:1]: We are already doing this in the golang workspace. Adding one more file is a small overhead, which in my opinion is worth it. The reason for this is, we are not expecting to have large number of Golang based directories, there will only be handful as we are building smaller RP compared to ARO-RP.

The discussion in the beginning of the project was that the go modules, should be treated as a separate entities that can be deployed and tested independently. 

The matrix style allows us to treat each module exactly like that, as an independent module, that is separated, linted and tested independently. This behavior is clearly signalling to the contributors how we are treating each module and how they should design their code. 

This is my thinking for bringing up the matrix style testing. Set the clear intentions and expectations for the code.

File: .github/workflows/ci-go.yml [hunk 5]
```diff
@@ -23,12 +30,20 @@ jobs:
         with:
           go-version-file: 'go.work'
           check-latest: true
+          cache: false
 
       - name: 'Test'
-        run: make test
+        uses: robherley/go-test-action@v0.2.0
```
Comment (s-amann) [inline .github/workflows/ci-go.yml:36]: what was wrong with the current strategy using the make target?

Comment (petrkotas) [reply .github/workflows/ci-go.yml:36]: Nothing, this one does exactly the same plus the auto-generated summary which can be seen here: https://github.com/Azure/ARO-HCP/actions/runs/8418935166/attempts/1#summary-23050469359 
This add a bit of convenience when looking at test results and figuring out what is wrong. It is my two cents to add convenience compared to grepping the logs for error.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (petrkotas) [reply .github/PULL_REQUEST_TEMPLATE.md:35]: Not now, which means it can be added later on.

File: .github/workflows/ci-go.yml
Comment (petrkotas) [reply .github/workflows/ci-go.yml:1]: I have removed the changes to make review easier.

File: .github/workflows/ci-go.yml (hunk 5 above)
Comment (petrkotas) [reply .github/workflows/ci-go.yml]: I removed the changes to make review easier.

**Review by mjlshen — APPROVED** [review]
/lgtm

Let's try it an iterate. Thanks!

**Pull request merged**